syntax = "proto3";
package mesh;

enum EnvelopeType {
  // Used mainly for hop-to-hop related info
  HANDSHAKE = 0;
  HEARTBEAT = 1;

  // Used for data transmission
  DATA = 2;
}

message PeerIP {
  string ip = 1;
  int32 tcp_port = 2;
  int32 udp_port = 3;
}

// Right now, can get this information from the remote address
// of the boost socket. In the future, this may be useful for
// implementing things like distributed hash tables or node discovery.
message PeerRecord {
  string peer_id = 1;
  PeerIP peer_ip = 2;
}

message Handshake {
  string proto_version = 1;
  PeerRecord peer_record = 2;
}


message Envelope {
  bytes msg_id = 1; // 16 raw bytes representing the UUID
  PeerIP from = 2; // ip and port
  PeerIP to = 3; // ip and port of remote or empty represents a broadcast message
  EnvelopeType type = 4;

  // For handshakes, it contains serialized Handshake
  // For everything else, it contains serialized RoutedPacket
  bytes payload = 5; // Serialized message
  bool expect_response = 6;
}
